services/src/main/scala/cromwell/services/IoActorRequester.scala

```diff
@@ -0,0 +1,46 @@
+package cromwell.services
+
+import akka.actor.{Actor, ActorRef}
+import akka.pattern.ask
+import akka.util.Timeout
+import com.typesafe.scalalogging.StrictLogging
+import cromwell.services.ServiceRegistryActor.{IoActorRef, NoIoActorRefAvailable, RequestIoActorRef}
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+
+trait IoActorRequester extends StrictLogging { this: Actor =>
+
+  val serviceRegistryActor: ActorRef
+
+  var _ioActorPromise: Option[Promise[ActorRef]] = None
+
+  def requestIoActor(backoffInterval: FiniteDuration = 1.minute): Future[ActorRef] = _ioActorPromise match {
+    case Some(promise) => promise.future
+    case None =>
+      val newPromise = Promise[ActorRef]
+      _ioActorPromise = Option(newPromise)
+      requestIoActorInner(newPromise, backoffInterval)
+      newPromise.future
+  }
+
+  private def requestIoActorInner(promise: Promise[ActorRef], backoffInterval: FiniteDuration): Unit = {
+    implicit val ec: ExecutionContext = context.system.dispatcher
+    implicit val timeout: Timeout = new Timeout(1.minute)
+
+    serviceRegistryActor ? RequestIoActorRef onComplete {
+      case Success(IoActorRef(actorRef)) =>
+        promise.complete(Success(actorRef))
+      case Success(NoIoActorRefAvailable) =>
+        logger.warn(s"No IoActorRef available for ${self.path} yet. Retrying in $backoffInterval.")
```

Comment:
Out of curiosity, and this could be a stupid question, is there a chance that IOActor might never be available and this will keep retrying infinitely? Although in that situation other parts of system might also fail no?

Reply:
Yup - in that case I guess we'd have to notice either the long stream of "no IoActorRef available" log messages, or that the metrics for successful streams was stuck at 0.

Either way, I think because archiving is a background task that's happening at least a year after someone's workflow has finished, the danger in this situation if it ever did happen is pretty small IMO